module-system/sov-state/src/codec/mod.rs

```diff
@@ -42,3 +43,38 @@ pub trait StateValueCodec<V> {
             .unwrap()
     }
 }
+
+pub trait EncodeLike<Ref: ?Sized, Target> {
+    fn encode_like(&self, borrowed: &Ref) -> Vec<u8>;
+}
+
+// All items can be encoded like themselves
+impl<C, T> EncodeLike<T, T> for C
+where
+    C: StateValueCodec<T>,
+{
+    fn encode_like(&self, borrowed: &T) -> Vec<u8> {
+        self.encode_value(borrowed)
+    }
+}
+
+// In borsh, a slice is encoded the same way as a vector
+impl<T> EncodeLike<[T], Vec<T>> for BorshCodec
+where
+    T: BorshSerialize,
+{
+    fn encode_like(&self, borrowed: &[T]) -> Vec<u8> {
+        borrowed.try_to_vec().unwrap()
```

Comment:
I'm a bit uncomfortable with implicit unwrap. I understand that is a restriction from borsh (i.e. their serialization is fallible, when it probably shouldn't be as there is no I/O), but then a type implementation that for some reason could be fallible would cause runtime panic.

Should we consider, in the future, wrapping the borsh traits? Something like

```rust
pub trait Serializable: BorshSerializable {
  fn to_vec(&self) -> Vec<u8> {
    borrowed.try_to_vec().unwrap()
  }
}
```

This way, we put everything into the same place so we make it easier to maintain in the future as if we ever have to change that, we just change the trait.

Reply:
I think borsh fixed this in the 1.0 release. We'll take care of these when we migrate.